### PR TITLE
Warning for special type of accounts

### PR DESCRIPTION
**Pull request by lnzainn**
#8440

### CODE REVIEW

**Comment by lnzainn** [comment]
@danieljames-dj @dunkOnIT good to merge this?

**Comment by danieljames-dj** [comment]
Thanks @lnzainn for raising this PR. But I really don't think this change aligns completely with what you intended for. By reading the issue that you tagged, I see that "all staff members who have permission to view [...]", but if your change is merged, this will shown to all users who have access to edit the user's account details which include the user themselves as well irrespective of whether they are staff or not).

(in case you are wondering if the user has access to edit their own account details then can they edit the WCA ID as well, the answer is no because WCA ID is added as editable field [only for this set of users](https://github.com/thewca/worldcubeassociation.org/blob/da0a9bdb2829bd853137a902871f33063923d19e/app/models/user.rb#L996).

My suggestion would be to add the same check in `app/views/users/edit.html.erb` as well.

But I also feel that it will be code duplication because you are writing `admin? || any_kind_of_delegate? || results_team? || communication_team?` in multiple places. I would also suggest to move this to a new method named something like `can_edit_any_users?` and use that method in both `app/views/users/edit.html.erb` and inside `editable_competitor_info_fields` method of user.rb.

**Comment by lnzainn** [comment]
Thanks for your suggestion, Daniel. This was something I discussed with Duncan and I get your point. On second thoughts, I agree, only those who can edit should be able to see that message. I'll propose the new changes

**Comment by dunkOnIT** [comment]
My first response was "Best to ask Daniel about this", I should have just stuck with that :laughing:

**Comment by lnzainn** [comment]
I wasn't sure if Board comes under 'admin?' method so I removed results_team? and replaced it with 'can_admin_results?' which I believe contains both board and WRT.

**Comment by danieljames-dj** [comment]
> I wasn't sure if Board comes under 'admin?' method so I removed results_team? and replaced it with 'can_admin_results?' which I believe contains both board and WRT.

Is there any reason why we want Board as well? Usually they won't handle these cases unless they are a delegate as well. If they have any case like this, it will be better if they do it through a local delegate or the necessary team because the chances are less that a non-Delegate Board member knows the correct procedur (they don't need to know as well because it's not part of Board workflows).

File: app/models/user.rb
Comment (danieljames-dj) [inline app/models/user.rb:1000]: Can you please use this method in line 991 as well? Something like `user == self || can_edit_any_user?`

**Comment by lnzainn** [comment]
Thanks Daniel for the catch, I have made the necessary changes.

**Review by danieljames-dj — APPROVED** [review]
Thanks @lnzainn, LGTM.

File: app/models/user.rb
Comment (dunkOnIT) [inline app/models/user.rb:718]: @danieljames-dj should `communication_team` have edit permissions in your view? In my mind, they should only need view permissions.

File: app/models/user.rb
Comment (dunkOnIT) [inline app/models/user.rb:991]: Bit weird to have an unless immediately after an if block - I suggest merging them so that the line is:

```ruby
if user == self || can_edit_any_user? || !cannot_edit_data_reason_html(user)
```

**Review by dunkOnIT — CHANGES_REQUESTED** [review]

File: app/models/user.rb
Comment (lnzainn) [reply app/models/user.rb:718]: @dunkOnIT WCT has been regularly handling wca id re assigning tasks which are trivial and can be performed by WRT too. We in fact had a discussion with what should be done but no conclusive solutions. @danieljames-dj should there be another WRT discussion on this? Perhaps if both teams vote together on what should be done, then we maybe will get a permanent solution

Comment (danieljames-dj) [reply app/models/user.rb:718]: @dunkOnIT Yes, currently, they handle requests like "change WCA ID to new account" themselves instead of forwarding them to WRT.

Comment (danieljames-dj) [reply app/models/user.rb:718]: @lnzainn We can think of having another discussion for sure. But I guess that won't block this PR to be submitted anyway because this PR doesn't change any permissions related stuffs. :-)

Comment (dunkOnIT) [reply app/models/user.rb:718]: Ah ok I see - and is this preferred by WRT, or is it worth re-evaluating if this is the best arrangement? (I assume it's preferred by WRT, but sometimes things can just be the way they are because no one's thought about them in a while - eg, WRC having permission to view registrations :P

Comment (lnzainn) [reply app/models/user.rb:718]: Not speaking of behalf of WRT, but when someone from wct made a rookie mistake of forwarding such request to wrt, then response from wrt was 'this can be done by wct themselves, wrt intervention is not required' 

Btw I'm that rookie who made this mistake when I had newly joined the team 😂

File: app/models/user.rb
Comment (dunkOnIT) [reply app/models/user.rb:991]: What do you mean by that?

**Review by gregorbg — APPROVED** [review]

File: app/models/user.rb
Comment (gregorbg) [reply app/models/user.rb:991]: Daniel is right, you cannot merge anything here. The GitHub diff view might be misleading: There's actually stuff happening inside the outer `if ...` block but _outside_ of the nested `unless` block.

Comment (dunkOnIT) [reply app/models/user.rb:991]: Ah sorry I see now! Probably just me misreading the diff.

**Review by dunkOnIT — APPROVED** [review]
LGTM!

**Pull request merged**
